Dashboard v2: Account Profile - Update username flow

pandah3 opened the pull request:
Linear: https://linear.app/a8c/issue/DOTDASH-513/implement-profile-update-username-flow

## Proposed Changes
Implementing the flow to update your username in the v2 Hosting Dashboard - Profile page.
### Before
How it looks on v1: https://wordpress.com/me/account

https://github.com/user-attachments/assets/34e75f30-634c-44c4-bf80-d0bbdb01e71a


### After (v2)

https://github.com/user-attachments/assets/32e1f593-5d2a-42b6-a7fd-d36ad4911370

There are no design mockups for the update username flow so I based the changes off of v1, having a new input field and radio options appear to confirm the username and optionally create a blog.

On v1, the new fields completely replace the existing form, but I just had them appear inline since this seems more accessible. I added lots of a11y helpers/attributes as well since things "appearing" on the screen without the user having taken any action can be challenging for a11y.

`/profile-gravatar` related changes are just around moving the mocks from `/profile-gravatar/test/__mocks__` to `/profile/__mocks__` since we also want to use them for the personal details section

**Note about the CSS addition:**
I explored various ways we could add the input validations:
- The custom `Edit` property on DataForms wasn't playing nice with state changes and causing focus issues so I couldn't use the DataViews validation.
- I stayed away from using what was used in v1 (`FormInputValidation` from `@automattic/components`).
- I attempted using `ValidatedTextControl` (internal component), but we have to bump up the version for `@wordpress/private-apis` to use it, which could lead to things potentially breaking. I moved forward with a different solution for now, but happy to try and bump it and switch to using `ValidatedTextControl`.
- I settled on using the `help` prop on `InputControl` and added CSS to style it, since `help` only adds the helper message.

## Why are these changes being made?

* To have feature parity with v1 so users can update their username in v2!

Note: This PR only covers the `Username` field and the flow to update the username in the Profile page, "Personal details" section. The email verification flow to update your email will be tackled in a separate [ticket](https://linear.app/a8c/issue/DOTDASH-514/implement-profile-update-email-flow)/PR to keep this PR contained to the username logic and not have it get bigger than it is now.

## Testing Instructions

1. `yarn start`
2. Visit http://calypso.localhost:3000/v2/me/profile
3. Assuming you're signed in as your a8c user, make sure the username field is disabled and you're unable to update your username.
4. In an incognito window or a different browser, login in to production wordpress.com as a non-A8C user (create a new user if you don't have one).
5. Then in the same browser/window, log in to local WP.com - http://calypso.localhost:3000/v2/me/profile. If you get stuck on a loop being directed back to the Calypso docs page and can't log in, try toggling your dev debug bar in the lower right corner: set Auth -> Oauth instead of cookies. You need to be logged into prod first as the non-A8C user for your auth to work locally as the same user.
6. Ensure that the username field isn't disabled and you can update it. Try updating it. You should see more fields appear. Make sure the validations for the inputs work.
8. Select yes for the blog radio option
9. Click submit and make sure you see a confirmation modal.
10. Click submit and ensure your username has been updated! Go to http://calypso.localhost:3000/v2/sites and make sure you see your new site with your new username as the subdomain.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

pandah3 committed:
Update developer opt-in description text for clarity

pandah3 committed:
Implement ability to change your username in the Hosting Dashboard:
- Port over helper functions from v1
- Add additional input field for username confirmation that appears on edit
- Add radio options to provision blogs with the new username
- Implement input validation and inline helper descriptions
- Add confirmation dialog with a warning about username changes
- Don't allow A12s to change their username; also don't allow if the user hasn't verified their email
- Handle canceling the username change
- Add success and error notifications
- Fix lint errors

pandah3 committed:
Refactor profile-personal-details file into focused components & utils to make it easier to read and test, and for separation of concerns:
- Pull username update form into its own dedicated UsernameUpdateForm component/file
- Move validation helpers into utils module
- Extract UsernameUpdateConfirmationModal into its own component/file
- Group files related to updating the username in its own folder

pandah3 committed:
Replace old design components with new ones and update business logic to be clearer:
- Use the help prop on InputControl to display helper text instead of using FormInputValidation. Add custom CSS to style the input field on error.
- Refactor validation factory to a more direct function approach to make username validation logic simpler and get rid of over-engineered solution
- Use debounce function from @wordpress/compose instead of a custom one
- Replace manual radio inputs with RadioControl component
- Replace divs with VStack

pandah3 committed:
Fix focus issue when typing in the username or username confirmation so you can type more than one letter at a time. DataForm recreates its custom Edit functions on every render so the fields keep losing focus when typing triggers state changes (e.g. bc of input validation).

Separate the username + username confirmation + email fields from DataForm and render them directly since they rely on more complex validation.

pandah3 committed:
Fix helper text and validation styling to show properly on error and success inputs

pandah3 committed:
Make everything accessible and user the proper components:
- Add ids, aria-labelledby, aria-describedby, etc. for better form structure, inputs, and modal a11y
- Add screen reader announcements for notices
- Replace <p>, <span> with the v2 Dashboard Text component; replace <div> with VStack for better layout and consistency across the dashboard

pandah3 committed:
Pass in role and aria-live props to make the Notice more accessible

pandah3 committed:
Add tests for personal details section, updating a username flow, confirmation dialog, and validation utils. Test for business logic, user flows, a11y, etc.

pandah3 committed:
Reinstate accidentally deleted file

pandah3 committed:
Move username validation API calls to @automattic/api-core and @automattic/api-queries for centralized API management:
- Add types, API calls to /api-core; add username change mutation to /api-queries
- Update components to use the React Query mutation

pandah3 committed:
Clean up: Fix lint errors, fix tests, add query mocks, condense files:
- Add missing mocks for getAllowedActions and submitUsernameChange
- Update renderWithUserData to properly mock the query functions with actual data for useSuspenseQuery
- Add missing api-queries mocks
- Condense usersettings mock to a single mock to share between profile sections (gravatar, personal details, etc)

pandah3 committed:
Fix lint errors

pandah3 committed:
Revamp feature for improvements and bug fixes:
- Extract username logic into separate component to keep the parent file focused on the overall profile details section
- Move form submission handlers from input component to form component for better separation of concerns
- Fix infinite API call loop
- Replace inline Notice components with WordPress snackbar notifications for consistency. Make sure snackbar is accessible.
- Update tests to reflect new component structure and removed props

pandah3 committed:
Add accessibility util to make sure that username update form popping up on username change is accessible

pandah3 committed:
Fix failing tests

pandah3 committed:
Fix lint errors and merge conflicts

pandah3 committed:
Fix bug that allowed submit before the username change validation was complete

pandah3 committed:
Remove changes that aren't needed anymore

fushar commented on packages/api-core/src/me-settings/mutators.ts:
```diff
@@ -37,3 +37,14 @@ export async function updateUserSettings(
 	return await wpcom.req.post( '/me/settings', payload );
 }
+
+export async function validateUsername( username: string ): Promise< UsernameValidationResult > {
```
Based on the endpoint path, these new functions should live in `/me-username`. See: pgz0xU-df-p2

pandah3 replied:
Ah yes, thank you!! Updated in [83747c2](https://github.com/Automattic/wp-calypso/pull/105912/commits/83747c26c9653ff4fe317b28a23373121a14ce6d)!

fushar commented on packages/api-queries/src/me-settings.ts:
```diff
@@ -22,3 +22,13 @@ export const userSettingsMutation = () =>
 		},
 	} );
+
+export const usernameChangeMutation = () =>
```
Similarly for the queries.

pandah3 replied:
Updated in [83747c2](https://github.com/Automattic/wp-calypso/pull/105912/commits/83747c26c9653ff4fe317b28a23373121a14ce6d)!

fushar commented on client/dashboard/me/profile-personal-details/index.tsx:
```diff
@@ -1,61 +1,139 @@
+		mutation.mutate( submissionEdits, {
+			onSuccess: () => {
+				setEdits( {} );
+				createSuccessNotice( __( 'Settings saved successfully.' ), {
```
We can use this newer mechanism using `meta`: https://github.com/Automattic/wp-calypso/pull/105880

pandah3 replied:
Updated in [9cf55c1](https://github.com/Automattic/wp-calypso/pull/105912/commits/9cf55c199b8ef078720d5615e50fbc63e31524ef)! I noticed the newer mechanism didn't support the other props we needed for a11y (`speak`, `politeness`, etc.), so I tweaked the snackbar meta logic to pass in additional props!

fushar commented on client/dashboard/me/profile-personal-details/index.tsx:
```diff
@@ -68,84 +146,89 @@ const fields: Field< UserSettings >[] = [
-		return data?.[ key as keyof UserSettings ] !== serverProfile?.[ key as keyof UserSettings ];
-	} );
+	return (
+		<>
```
Do we need this outer fragment?

pandah3 replied:
Removed!!

fushar commented on client/dashboard/me/profile-personal-details/index.tsx:
```diff
@@ -1,61 +1,139 @@
 		label: __( 'I am a developer' ),
 		type: 'boolean',
-		description: __( 'Opt me into previews of new developer-focused features.' ),
+		description: __( 'Opt in to previews of new developer-focused features.' ),
```
Is this change intended?

pandah3 replied:
Yes, it was included in the Figma mockup! QOBjujZah0UlGDDdLKZhzi-fi-8987-122870
I originally intended for this PR to address all the fields of the "Personal details" section, but the username part got too big 😅  It's a little stray change that got included 🙈

fushar commented on client/dashboard/me/profile-personal-details/update-username/index.tsx:
```diff
@@ -0,0 +1,168 @@
+				</VStack>
+			) }
+
+			<HStack justify="flex-start">
```
Let's use the new `<ButtonStack>` for this.

pandah3 replied:
Updated in [8058dee](https://github.com/Automattic/wp-calypso/pull/105912/commits/8058dee685c52220059f4da74df63e0c2df54705), thank you!!

pandah3 committed:
Use new ButtonStack component as the button wrapper for spacing consistency

pandah3 committed:
Remove unnecessary outer fragment

pandah3 committed:
Pass in additional props to the snackbar meta so we can use a11y controls in the notice. Update the profile settings notice to use the new meta.snackbar pattern and update the copy to match the new guidelines.

pandah3 committed:
Refactor logic in api-core and api-queries to organize username functionality by endpoint structure:
- Move username-related functions and types to /me-username to match the /me/username endpoint. Keep UserSettings in /me-settings since it's shared across features.
- Move mutation to new /me-username file to match the api-core structure

pandah3 committed:
Update tests

pandah3 committed:
Fix TypeScript import error for DataPointDate from @automattic/charts

pandah3 committed:
Merge branch 'trunk' into update/profile-personal-details

fushar commented on client/dashboard/me/profile-personal-details/update-username/confirmation-modal.tsx:
```diff
@@ -0,0 +1,57 @@
+				aria-describedby="username-change-description"
+				data-testid="username-change-dialog"
+			>
+				<h3 id="username-change-title">{ __( 'Confirm username change' ) }</h3>
```
This is not a canonical way to add heading to modals; the styles are different from the rest of the product.

Unfortunately `ConfirmDialog` does not support titles, so please use this one instead: [ConfirmModal](https://github.com/Automattic/wp-calypso/tree/25ad32b94e86e2bbbc84c77842407ce2328229fa/client/dashboard/components/confirm-modal).

pandah3 replied:
Oops, sorry, I forgot to use `<Heading>` and apply the right styling from Figma to match (I referenced a different dialog since there wasn't a mockup for this one). Switched to use `ConfirmModal` in [29c3c63](https://github.com/Automattic/wp-calypso/pull/105912/commits/29c3c6366f3a1693b43fc99a614f7a4d73d4ce8e)!

Side note: I wonder if we could consolidate `ConfirmModal` and `ConfirmDialog` into a single component. Since they both address confirmation workflows and they're both built on top of `Modal`, it feels like we have competing components that could create confusion about which one to use when.

Maybe we could adjust `ConfirmDialog` to support the additional features that `ConfirmModal` currently offers (button states, header customization, etc.). This would give us a more consistent approach across the board as well since some components are currently using `ConfirmModal`, while others are using `ConfirmDialog`.

fushar commented on packages/api-queries/src/me-username.ts:
```diff
@@ -0,0 +1,13 @@
+			submitUsernameChange( username, action ),
+		onSuccess: () => {
+			// Invalidate user settings to refetch updated data
+			queryClient.invalidateQueries( { queryKey: [ 'me', 'settings' ] } );
```
We can simply use this to avoid having to type the key


```suggestion
			queryClient.invalidateQueries( userSettingsQuery() );
```

pandah3 replied:
Updated in [4845b6b](https://github.com/Automattic/wp-calypso/pull/105912/commits/4845b6bdb9400c1ef0fad49c0f9e4460b29eae51), thank you!

fushar commented on packages/api-core/src/me-username/mutators.ts:
```diff
@@ -0,0 +1,13 @@
+	return await wpcom.req.get( `/me/username/validate/${ username }` );
+}
+
+export async function submitUsernameChange(
```
[nit] I think it can be simply `updateUsername()` or `changeUsername()`.

pandah3 replied:
Updated to `updateUsername()` in [4845b6b](https://github.com/Automattic/wp-calypso/pull/105912/commits/4845b6bdb9400c1ef0fad49c0f9e4460b29eae51)!

fushar commented on client/dashboard/me/profile-personal-details/username-section.tsx:
```diff
@@ -0,0 +1,231 @@
+	const [ validationResult, setValidationResult ] = useState< ValidationResult | null >( null );
+
+	const usernameMutation = useMutation( usernameChangeMutation() );
+	const confirmFormRef = useRef< HTMLDivElement >( null );
```
What is this ref for?

pandah3 replied:
It's to help with focus for a11y! I added it initially, but realized it might be more disruptive moving the focus to the new input field if they're still typing in the old one, so I hadn't fully implemented it. Removed it for now!

fushar commented on client/dashboard/me/profile-personal-details/update-username/username-validation-utils.ts:
```diff
@@ -0,0 +1,82 @@
+	return validationResult.allowed_actions ?? {};
+}
+
+export async function submitUsernameChange( username: string, action: string ) {
```
Is this actually being used?

pandah3 replied:
Ah no, removed in [4845b6b](https://github.com/Automattic/wp-calypso/pull/105912/commits/4845b6bdb9400c1ef0fad49c0f9e4460b29eae51). Thank you for catching!

p-jackson commented on client/dashboard/me/profile-personal-details/index.tsx:
```diff
@@ -1,61 +1,133 @@
+	const canChangeUsername = userSettings?.user_login_can_be_changed ?? true;
+
+	useEffect( () => {
+		const params = new URLSearchParams( window.location.search );
+		if ( params.get( 'usernameChangeSuccess' ) === 'true' ) {
```
So great that we're thinking about `aria-live`! It's great that HD is going to have a much better a11y story than v1 Calypso.

For this particular "show message on load" behaviour we already have a `<FlashMessage>` component. Can we bake the `aria-live` functionality into `FlashMessage` instead of implementing it in this one component?

pandah3 replied:
Woohoo, glad to hear that a11y is a part of HD efforts!!

`<FlashMessage>` is exactly what I needed! Replaced the current logic with it in [4f3f89a](https://github.com/Automattic/wp-calypso/pull/105912/commits/4f3f89aa4e7c851dd3d0da5dc9a05cf52c7462c1). I'll address the `aria-live` functionality in a separate PR as Ashar recommended [below](https://github.com/Automattic/wp-calypso/pull/105912/#discussion_r2380794017)!

I found a bug in `<FlashMessage>` that I snuck in a small fix for in [a64d909](https://github.com/Automattic/wp-calypso/pull/105912/commits/a64d90997a926a12ae9e48e8eb1f3f6ab7b743f7). `key` is a [reserved prop](https://reactjs.org/link/special-props) that we can't use so I changed it to `queryParam`. I was getting an error since I needed to pass a different query param than the default (I since changed mine to align with the what other features were doing (`update=<setting_name>`)).

```
Warning: FlashMessage: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. 
```

p-jackson replied:
> `key` is a [reserved prop](https://reactjs.org/link/special-props) that we can't use so I changed it to `queryParam`

Yes that  `key` prop was an oversight on my part! It was also fixed in #105695, along with other `<FlashMessage>` improvements.

fushar commented on client/dashboard/me/profile-personal-details/update-username/index.tsx:
```diff
@@ -0,0 +1,168 @@
+
+	useEffect( () => {
+		if ( showRadioOptions && ! hasAnnouncedRadioOptions ) {
+			announceToScreenReader(
```
Hi @pandah3, thanks for also considering a11y in this PR. Very much appreciated!

I think this is an opportunity to share this to everyone, as we currently almost don't use a11y-related logic in the HD codebase, other than the ones provided by the Core components out of the box. I think this pattern is applicable in many screens as well.

As this PR is introducing a new pattern in HD, could we please:

- Post a P2 in the hosting dashboard P2, sharing how you think we should approach a11y in many parts of the HD, and
- Separate this particular a11y logic in a new PR.

HD is being developed by many devs across the company. By doing that, the idea is to get collaboration, share ideas, and prevent devs from (mistakenly) blindly copy-pasting logic without actually understanding why (which has happened in v1).

Many thanks!

pandah3 replied:
Yes, that's a great idea 🙏 Thank you for the suggestion!!

I removed the a11y logic for now and will open up a separate PR for it + get the discussion started on an a11y approach! Sharing more broadly and being aligned across the board sounds right in line with what we're trying to achieve!!

Thank you for the multiple PR reviews as well 🙇‍♀️ !!

fushar commented on client/dashboard/app/snackbars/index.tsx:
```diff
@@ -20,6 +20,9 @@ declare module '@tanstack/react-query' {
 			snackbar?: {
 				success?: string;
 				error?: string;
+				speak?: boolean; // announce to screen readers
```
This is related to my previous comment. We should definitely P2 about this and ask about designer's input, so let's also separate this to another PR.

A good outcome is maybe to just add these as default values to the snackbar logic, so that devs don't have to think what politeness they should set in their snackbar :)

pandah3 replied:
That makes sense! It would make things a lot easier if this was handled behind the scenes so a11y is just baked right into everything! I removed the fields and I'll include this in the P2 as well!

pandah3 committed:
Switch to use <ConfirmModal> for the update username confirmation modal for consistency. Pass in loading state for button

pandah3 committed:
Fix element and variable substitution in the translation string

pandah3 committed:
Improvements and code cleanup for clarity:
- Use userSettingsQuery() for query invalidation instead of hardcoded key
- Update username update mutator name for clarity. Update mutation name in api-queries to match. Update all imports and references.
- Remove unused username validation util

pandah3 committed:
Extract a11y-related changes to a separate PR

pandah3 committed:
Use <FlashMessage> instead of custom logic to show a success message once the update flow is complete

pandah3 committed:
Reassign query param and value upon successful username update so it aligns with other query param updates

pandah3 committed:
Replace 'key' prop with 'queryParam' in <FlashMessage> since 'key' is a reserved prop and can't be passed to components

pandah3 committed:
Fix tests

pandah3 committed:
Remove a11y tests

pandah3 committed:
Merge branch 'trunk' into update/profile-personal-details

abdullah-kasim commented on client/dashboard/me/profile-personal-details/test/index.test.tsx:
```diff
@@ -0,0 +1,316 @@
+	return result;
+};
+
+describe( 'PersonalDetailsSection', () => {
+	beforeEach( () => {
+		jest.clearAllMocks();
```
@p-jackson I'd love to have your input on the tests - you've previously offered [some suggestions on a previous PR](https://github.com/Automattic/wp-calypso/pull/105389#discussion_r2314922499) of mine that I've since integrated in my code writing (and has reduced the chore-ness, and increased the average value of the tests that I wrote!)

As it stands, I can see that we can trim the test to only the essential cases -
Tests that I think are essential:
* Username editing logic (and disablement for A8c users, and so on).
* Save button submission logic.
* Accessibility logic.

Tests that I don't think are essential:
* Testing if form fields are filled properly - I think this is a very common pattern throughout the codebase that we might just be testing the library in that case? And theoretically, save button submission logic test should test this case for us.
* Clicking developer checkbox.

What do you ( @pandah3 and @p-jackson ) think of dropping the tests that I think are non-essential?

pandah3 replied:
I like that!! I would've said testing the developer checkbox is important because it's its own mini "section", but you're right, the tests we have for it now aren't doing anything substantial.

I do think the `Basic rendering - renders the form with all sections` test is good to keep, just to make sure the UI looks as how we expect.

I'll remove the "testing if form fields are filled properly" + developer checkbox tests. Thanks, Abdullah 🙏

pandah3 replied:
Removed them in [61fcd78](https://github.com/Automattic/wp-calypso/pull/105912/commits/61fcd785193e05b6f30a4c65161a8ab1b1c88834)!

abdullah-kasim commented on client/dashboard/me/profile-personal-details/test/index.test.tsx:
```diff
@@ -0,0 +1,316 @@
+			} );
+		} );
+
+		it( 'shows error notice on mutation error', async () => {
```
I think this test is mnissing a `mockCreateErrorNotice` assertion that an error has been shown on the snackbar?

pandah3 replied:
Oh yeah, good catch. I think I had this as a dirty state test or something initially.
I was handling the mutation error differently than the success case, but I'll standardize it so it's the same experience.

pandah3 replied:
Fixed in [89d6add](https://github.com/Automattic/wp-calypso/pull/105912/commits/89d6add896462cbd656cefb26853495a013649a3)!

abdullah-kasim reviewed (commented):
Did a quick, partial review. I'll try and review more later. Thanks for all the work!

pandah3 committed:
Show error message via snackbar notice when the updateUsername mutation fails, just like we do for success cases. Add tests to mock the mutation fail request

pandah3 committed:
Remove non-essential tests

pandah3 committed:
Merge branch 'trunk' into update/profile-personal-details

pandah3 committed:
Merge branch 'trunk' into update/profile-personal-details

brunobasto commented on client/dashboard/me/profile-personal-details/update-username/confirmation-modal.tsx:
```diff
@@ -0,0 +1,52 @@
+	isBusy,
+}: UsernameUpdateConfirmationModalProps ) {
+	if ( ! isOpen ) {
+		return null;
```
@pandah3  I might have done this myself, but I forgot. Is this early `return null` necessary if the <ConfirmModal>` component already receives a `isOpen` prop?

pandah3 replied:
I think the `ConfirmModal` component would still get created even when it doesn't need to if we didn't return `null` when `isOpen` is false.

brunobasto commented on client/dashboard/me/profile-personal-details/update-username/test/confirmation-modal.test.tsx:
```diff
@@ -0,0 +1,70 @@
+	} );
+
+	describe( 'Conditional rendering and content', () => {
+		it( 'renders when isOpen is true', () => {
```
If the `isOpen` behavior is controlled by `<ConfirmModal>` we might be able to remove these first two tests since we'd be testing an external component.

pandah3 replied:
True. I think the second one makes more sense to keep since we're testing our `return null` logic. But it wouldn't really make sense to have one but not the other. I don't think it matters either way because they're pretty trivial tests. Removed in [91fedfb](https://github.com/Automattic/wp-calypso/pull/105912/commits/91fedfb3cb0898ba72248840871f8f473418f06c)!

brunobasto commented on client/dashboard/me/profile-personal-details/update-username/username-validation-utils.ts:
```diff
@@ -0,0 +1,75 @@
+
+export type ValidationResult = UsernameValidationResult;
+
+async function validateUsernameInternal(
```
This could be a great candidate for an isolated function unit test

pandah3 replied:
I have some tests for it here: https://github.com/Automattic/wp-calypso/pull/105912/files#diff-922f5b2de5140fee04863e1cd275b4797cd9a88820dd18fc9d05ffd48ed53b59R114-R139

It's testing `validateUsernameDebounced._originalFn` which should be directly testing the validation logic (without the debounce) through the mock!

brunobasto replied:
Ah cool. I didn't catch that initially.

brunobasto reviewed (approved):
Tested and works as expected 🎉 

I love that you included comprehensive tests 🛡️ 

Left a few minor comments, but looks good to me.

pandah3 committed:
Remove unnecessary tests

pandah3 committed:
Merge branch 'trunk' into update/profile-personal-details
